src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,201 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Net.Http;
+using System.Threading.Channels;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+using Microsoft.AspNetCore.Server.Kestrel.Core.WebTransport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.WebTransport;
+
+/// <summary>
+/// Controls the WebTransport session and keeps track of all the streams.
+/// </summary>
+internal class WebTransportSession : IWebTransportSession
+{
+    private readonly CancellationTokenRegistration _connectionClosedRegistration;
+
+    // stores all created streams (pending or accepted)
+    private readonly ConcurrentDictionary<long, WebTransportStream> _openStreams = new();
+    // stores all pending streams that have not been accepted yet
+    private readonly Channel<WebTransportStream> _pendingStreams;
+
+    private readonly Http3Connection _connection;
+    private readonly Http3Stream _connectStream = default!;
+    private bool _isClosing;
+
+    private static readonly ReadOnlyMemory<byte> OutputStreamHeader = new(new byte[] {
```

Comment:
Use the `ReadOnlySpan<byte>` pattern we use in other places
https://github.com/dotnet/aspnetcore/blob/3ea008c80d5cc63de7f90ddfd6823b7b006251ff/src/Servers/Kestrel/Core/src/Internal/Http/DateHeaderValueManager.cs#L15-L16
(ignore the `u8` part since you can't use that here)

Reply:
I pass this into the `PipeWriter.WriteAsync` method, which takes a `ReadOnlyMemory`. Is there a way to efficiently convert from a span to the memory?

Reply:
> I pass this into the `PipeWriter.WriteAsync` method, which takes a `ReadOnlyMemory`.

👍 ok, just making sure. Fine to keep as `ReadOnlyMemory`.